codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -0,0 +1,467 @@
+        }
+        try {
+            boolean unexpected = false;
+            if (msg instanceof Http2StreamFrame) {
+                Object streamObject = ((Http2StreamFrame) msg).stream();
+                if (!(streamObject instanceof Http2StreamChannel)) {
+                    // TODO: swap to failing the promise. Today, however, the promise is likely
+                    // being ignored.
+                    throw new IllegalArgumentException(
+                        "Http2StreamFrame.stream() must be an Http2StreamChannel: " + streamObject);
+                }
+                int streamId = ((Http2StreamChannel) streamObject).stream.id();
+                if (msg instanceof Http2DataFrame) {
+                    Http2DataFrame frame = (Http2DataFrame) msg;
+                    http2Handler.encoder().writeData(http2HandlerCtx, streamId, frame.content().retain(),
+                        frame.padding(), frame.endStream(), promise);
+                } else if (msg instanceof Http2HeadersFrame) {
+                    Http2HeadersFrame frame = (Http2HeadersFrame) msg;
+                    http2Handler.encoder().writeHeaders(
+                        http2HandlerCtx, streamId, frame.headers(), frame.padding(), frame.endStream(), promise);
+                } else if (msg instanceof Http2ResetFrame) {
+                    Http2ResetFrame frame = (Http2ResetFrame) msg;
+                    http2Handler.resetStream(http2HandlerCtx, streamId, frame.errorCode(), promise);
+                } else {
+                    unexpected = true;
+                }
+            } else if (msg instanceof Http2GoAwayFrame) {
+                Http2GoAwayFrame frame = (Http2GoAwayFrame) msg;
+                int lastStreamId = http2Handler.connection().remote().lastStreamCreated()
+                    + frame.extraStreamIds() * 2;
+                http2Handler.goAway(
+                    http2HandlerCtx, lastStreamId, frame.errorCode(), frame.content().retain(), promise);
+            } else {
+                unexpected = true;
+            }
+            if (unexpected) {
+                // TODO: swap to failing the promise. Today, however, the promise is likely being
+                // ignored.
+                throw new IllegalStateException("Unexpected Http2Frame type: " + msg.getClass().getName());
+            }
+        } finally {
+            ReferenceCountUtil.release(msg);
+        }
+    }
+
+    Http2StreamChannel createStreamChannel(ChannelHandlerContext ctx, Http2Stream stream,
+            ChannelHandler handler) {
+        EventLoopGroup group = streamGroup != null ? streamGroup : ctx.channel().eventLoop();
+        Http2StreamChannel channel = new Http2StreamChannel(stream);
+        channel.pipeline().addLast(handler);
+        group.register(channel);
+        return channel;
+    }
+
+    private static ChannelHandlerInvoker invoker(ChannelHandlerContext ctx) {
+        return ((EventLoop) ctx.executor()).asInvoker();
```

Comment:
If it isn't an `EventLoop`, what will the code do? It seems the code would be forced to throw an exception.

This is a problem that scared me quite a bit when porting to 4.1 until I found the code you see here. In 5.0 there is `ChannelHandlerContext.invoker()`. But in 4.1 there is only `AbstractChannelHandlerContext.invoker()` and `AbstractChannelHandlerContext` and `DefaultChannelHandlerContext` are package-private.

There are three cases that matter for this code:

 1. The executor is the `Channel`'s. `Channel`s are required to use `EventLoops`, so there is no problem.
 2. An `EventExecutor` was specified to `ChannelPipeline` when registering the handler. `EventExecutor` is an `EventLoop`, so there is no problem.
 3. A `ChannenlHandlerInvoker` was specified to `ChannelPipeline` when registering the handler. `ctx.executor()` calls `ChannelHandlerInvoker.executor()`. But nothing requires the returned type to be `EventLoop` which means I'm not necessarily able to get the invoker. There is irony lurking here.

My preference would be to have `ChannelHandlerContext.invoker()` as in 5.0 :smile:, but that is either an interface change or making some internal classes public. But I also have very little problem ignoring the issue, because it seems like a severe edge case.

If you'd like, I can detect the broken case and throw an exception with a clear error message, but I honestly expect it wouldn't save anyone much time if they hit this problem.